Enhancements to 90 Day Uptime Visualization

OstlerDev opened the pull request:
# Overview
The "Last 90 day" view in my opinion could be further utilized to display the services status for the day. I have made some changes that I think look good while adding much more detail to the service's status for a particular day. The changes display if a service was online, or if there was an outage, how severe the outage was based on the number of bad data points we recieved. 

Currently the application will display a grey cell for no issues, and a bright red cell if there is even a single issue. In my opinion showing a server as "red" means it was completely offline or severely impaired, and missing a single ping shouldn't be cause for showing an "outage" for the full day.

It was that reason that I decided it might be nice if it was a little more dynamic.

## Changelog
### Fixed
- [DevOps] Development Dockerfile with hot reloading, since I needed it working to get this working haha.
- [UI] Fixed display of day (xAxis) on the Service Heatmap (on the service page), see picture below
- [Go] Calculation of Uptime when a service has never been online
### Changed
- [Env] Default to keep records for 365 days before purging
### Added
- [Go] Better/more diverse sample data for service hits/failures.
- [UI & Go] Added uptime calculation for the past 12 months
- [UI] **Dynamic Color Coding:** Modified the day cells to reflect a range of statuses/outage severities with a new color scheme. 
    - ![#e9e9e9](https://placehold.co/15x15/e9e9e9/e9e9e9.png) `#e9e9e9` Light Grey for days with no data
    - ![#4CAF50](https://placehold.co/15x15/4CAF50/4CAF50.png) `#4CAF50` Green for days with 100% uptime
    - A gradient from light green to yellow, orange, and red, indicating the severity of outages based on duration.
        - ![#98EE99](https://placehold.co/15x15/98EE99/98EE99.png) `#98EE99` (Minor outage) Light green for days with between 1-30 failures (up to 30 minutes of downtime)
        - ![#FFEB3B](https://placehold.co/15x15/FFEB3B/FFEB3B.png) `#FFEB3B` (Moderate outage) Yellow for days with between 31-120 failures (up to 2 hours of downtime)
        - ![#FF9800](https://placehold.co/15x15/FF9800/FF9800.png) `#FF9800` (Major outage) Orange for days with between 121-240 failures (up to 4 hours of downtime)
        - ![#F44336](https://placehold.co/15x15/F44336/F44336.png) `#F44336` (Critical outage) Red for days with over 240 failures (over 4 hours of downtime)
 
## Ideas for improvement
It would be great if the outage severities could be set inside the settings of Statping-ng, I just didn't have time at the moment to add it, but I did leave everything very configurable in the Vue file in case someone wants to get around to it eventually. If I get some time I will try to add it in another MR, but no promises.

## Preview
<img width="990" alt="status-graph" src="https://github.com/statping-ng/statping-ng/assets/1490586/95cb4107-183f-429e-9f11-7e30bea30141">
<img width="1106" alt="4-service-view" src="https://github.com/statping-ng/statping-ng/assets/1490586/4ad6b6fd-ed69-420e-94f6-18acbf158a02">
<img width="691" alt="6-service-view" src="https://github.com/statping-ng/statping-ng/assets/1490586/9253351f-410c-4b18-86f8-002ce4c77ac6">
<img width="979" alt="12-month-uptime" src="https://github.com/statping-ng/statping-ng/assets/1490586/cc38133b-679c-4251-ac1d-2722c19ba073">
<img width="983" alt="service-heatmap" src="https://github.com/statping-ng/statping-ng/assets/1490586/b43c5a5a-8b33-44f6-97eb-47755648a3a1">


## What it used to look like
<img width="971" alt="day-view-old" src="https://github.com/statping-ng/statping-ng/assets/1490586/e021a068-fd9a-4b74-8fdb-7678cc82aa20">
## Proposed Changes
<img width="990" alt="status-graph" src="https://github.com/statping-ng/statping-ng/assets/1490586/95cb4107-183f-429e-9f11-7e30bea30141">

OstlerDev committed:
Add dev getting started docs

OstlerDev committed:
Use correct docker image repository

OstlerDev committed:
UI: Display how severe an outage is in day view

OstlerDev committed:
Go: Add failure samples for outage severities

OstlerDev committed:
Docker: Fix Dev Dockerfile/Compose!

OstlerDev committed:
Show if service was online or offline in day cell

OstlerDev committed:
Go: Generate sample success hits going back up to 90 days

OstlerDev committed:
Add more details to dev readme

OstlerDev committed:
Cleanup GroupServiceFailures

jemand771 commented:
wow, this looks good! I agree that the thresholds _could_ be configurable, but this is already an improvement, so no worries :)
let's see if CI broke again

also, thanks for fixing the dev setup - I've been procrastinating on that for, what, _a year_ now?

OstlerDev commented:
Glad you like it! 

Happy to help haha, sometimes it's the small things that are just hard to get to!

Looks like I might have broken the tests with changing the sample data 😵 Though when I took a look at the failing tests it doesn't seem to be that 🤔

jsnbuchanan commented:
Love this PR. Thank you @OstlerDev!

OstlerDev commented:
I built and released a version on Docker Hub so that anybody who wants can test this out.

`ostlerdev/statping-ng:v0.91.0-beta4` [Link to Docker Hub](https://hub.docker.com/r/ostlerdev/statping-ng/tags)

Note: Docker building was broken, but I fixed it, and included this fix in this new Pull Request: https://github.com/statping-ng/statping-ng/pull/259

OstlerDev committed:
Merge branch 'statping-ng:dev' into ui-dynamic-daily-status-bar

jemand771 commented on types/hits/samples.go:
```diff
@@ -15,20 +15,51 @@ var SampleHits = 99900.
-			log.Error(err)
-			return err
-		}
+	records_1 := createHitsAt(1, -90 * types.Day, 1 * time.Hour)
```
aha, I went through the postman tests and figured out this is what's "wrong":

It wants at least 100 hits..
https://github.com/statping-ng/statping-ng/blob/4d872facdafa1f527d5b2c46e4c09623804f5e2c/dev/postman.json#L1479-L1481

..in the last three days..
https://github.com/statping-ng/statping-ng/blob/4d872facdafa1f527d5b2c46e4c09623804f5e2c/dev/postman.json#L20-L22

.. which, at 1/h, is not the case (that's where the actual 72 / expected 100 is coming from)

Since this is only checked for the service with id 1, it really doesn't matter where we/you change this -
* bump up the sample hit frequency here (to 42-ish minutes or lower for 100 in 3 days)
* reduce the expected number of hits to 72
* increase the time range the tests looks at (`{{start_time}}` is only used in this one test)

like I said, I don't mind any of those. you pick :b

OstlerDev replied:
I went through and added more data samples and it works consistently now :D Figured I would add more realistic data with minute intervals.

I was able to run all the tests in postman successfully locally, hopefully this works now 😸

OstlerDev committed:
Fix tests by generating more realistic samples

OstlerDev commented:
Alright, that should fix the issue 😸

OstlerDev committed:
Merge branch 'dev' into ui-dynamic-daily-status-bar

OstlerDev committed:
Fixup Service Heatmap on Service Page

OstlerDev committed:
Display Uptime % for past 12 Months

Keep last 365 days of records and fix issue with uptime calculation

OstlerDev committed:
Allow higher precision for past year uptime

OstlerDev commented:
I went through and took another quick look and decided to add another couple things to this pull request that I think enhance it a little further. Figured I would get to those while I was working on things. Let me know what you think of all the changes, I figure they are in line with the goal of this project so hopefully they are easy wins 😸 

1. Added Uptime calculation for the past year, to give people the ability to view their SLA Uptime over the past 12 months.
<img width="979" alt="Screenshot 2024-03-25 at 12 53 19 AM" src="https://github.com/statping-ng/statping-ng/assets/1490586/ff3088f7-8844-48a2-83cf-f4abb74845ac">

2. Fixed the Service Heatmap (on the service page) so that it properly displays the day of the month, along with a better tooltip
<img width="983" alt="Screenshot 2024-03-25 at 12 53 40 AM" src="https://github.com/statping-ng/statping-ng/assets/1490586/cecf6654-d2d2-4bd7-8af7-a15f6da4e0fd">

OstlerDev committed:
Try using Go 1.20.0 for running Tests

OstlerDev commented:
Looks like there was an issue with the testing workflows building the Go app properly. I saw that it might be an issue with newer versions, so I pinned the tests to use Go v1.20.0 to match the Main Dockerfile exactly. It looks like it was installing Go v1.20.14 I believe, so maybe using an older version will work? I will attempt to run the worlflows locally to see if I can get them working as expected.

OstlerDev committed:
Add GOPROXY directly?

OstlerDev committed:
Attempt to set goroot to path correctly

and also check the go version

OstlerDev committed:
Use Go v1.21.8 for running tests

OstlerDev commented:
Alright @jemand771, I finally got the tests working 😵 Turns out there was a couple errors in the test pipelines, but fixing a couple things with the Go version/setup seems to make them run just fine when I tested them on my personal GitHub repo 😸 

Hope they also work for you, they are running like a charm for me now 😄  (other than `docker-build` taking 1h+ to run haha)

adamboutcher commented:
Sorry for being silent, is this merge good to go?

jemand771 commented:
something's up with the tests (but only somtimes?) that I'd like to sort out first: #282

NikoCat233 commented:
![image](https://github.com/statping-ng/statping-ng/assets/139348239/5f16bef4-e1eb-4fbd-9915-09b6edc44643)
I can not run your pr

OstlerDev commented:
It looks like the tests are now properly running! Great job figuring that out! Figured I would give this a little bump now 😜 

I have been running this in two personal production instances for well over a year, haven't run into any issues yet! 🥳
